File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        is_balanced = True

        def dfs(root):
            if not root: 
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)

            nonlocal is_balanced
            is_balanced = is_balanced and abs(left - right) <= 1

            return 1 + max(left, right)
        
        dfs(root)
        
        return is_balanced

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.swift
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     var val: Int
 *     var left: TreeNode?
 *     var right: TreeNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func goodNodes(_ root: TreeNode?) -> Int {
        guard let root else { return 0 }
        func dfs(_ node: TreeNode?, _ prevMax: Int) -> Int {
            guard let node else { return 0 }

            var good: Int = if node.val >= prevMax { 1 } else { 0 }
            let newMax = max(prevMax, node.val)

            let leftCount = dfs(node.left, newMax)
            let rightCount = dfs(node.right, newMax)

            return good + leftCount + rightCount
        }

        return dfs(root, root.val)
    }
}
